### PR TITLE
Refactors xgbserver and kfserver to separate pypi folders

### PR DESCRIPTION
**Pull request by cliveseldon**
* Fix `kfserving`/`kfserver` name mismatch in setup.py. Which do we want?. Went with `kfserving`
 * Fix argument inheritance in argsparse for xgboostserver to kfserving server

A larger question. I'm not sure one is supposed to have multiple calls to `setup` in `setup.py`. Should we have 2 modules each with a `setup.py` in a subfolder? 
For example:
```
   python/kfserving/setup.py
   python/kfserving/kfservng/<python module code>
   python/xgbserver/setup.py
   python/xgbserver/xgbserver/<python module code>
```
 See: https://python-packaging.readthedocs.io/en/latest/minimal.html

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/63)

### CODE REVIEW

File: python/kfserving/server.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
 DEFAULT_GRPC_PORT = 8081
 DEFAULT_PROTOCOL = "tensorflow/HTTP"
 
-parser = argparse.ArgumentParser()
+parser = argparse.ArgumentParser(add_help=False)
```
Comment (ellis-bigelow) [inline python/kfserving/server.py]: Why did you change this?

File: python/kfserving/server.py [hunk 2]
```diff
@@ -53,7 +53,7 @@ def start(self, models=[]):
             (r"/models/([a-zA-Z0-9_-]+):predict",
              ModelPredictHandler, dict(models=self.registered_models)),
         ]))
-
```
Comment (ellis-bigelow) [inline python/kfserving/server.py]: Which linter are you using? Somehow my pylint didnt catch this.

File: python/setup.py [hunk 3]
```diff
@@ -6,11 +6,11 @@
     version='0.1.0',
     author_email='ellisbigelow@google.com',
     license='../../LICENSE.txt',
-    url='https://github.com/kubeflow/kfserving/python/kfserving/kfserver',
+    url='https://github.com/kubeflow/kfserving/python/kfserving/kfserving',
     description='Model Server for arbitrary python ML frameworks.',
     long_description=open('README.md').read(),
     python_requires='>3.4',
-    packages=find_packages("kfserver"),
+    packages=find_packages("kfserving"),
```
Comment (ellis-bigelow) [inline python/setup.py]: I agree with this naming convention :). I wonder if we can pull this out a level so that we don't have doubly nested kfserving.

I think the config file is completely fine for multiple setup calls -- at least it was working for me.

/python/setup.py
/python/kfserving/...
/python/xgbserver/...

File: python/xgbserver/__main__.py [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
 import argparse
 
-from model import XGBoostModel
+from xgbserver import XGBoostModel
 
 DEFAULT_MODEL_NAME = "default"
 DEFAULT_LOCAL_MODEL_DIR = "/tmp/model"
 
-parser = argparse.ArgumentParser()
+parser = argparse.ArgumentParser(parents=[kfserving.server.parser])
```
Comment (ellis-bigelow) [inline python/xgbserver/__main__.py]: I was getting the inheritance behavior without parents. I think this is fine, but I'm curious what issue you were running into.

Comment (cliveseldon) [reply python/xgbserver/__main__.py]: Running from the command line I wasn't getting the inheritance.

File: python/kfserving/server.py (hunk 2 above)
Comment (cliveseldon) [reply python/kfserving/server.py]: I'm using builtin with pyCharm

File: python/kfserving/server.py (hunk 1 above)
Comment (cliveseldon) [reply python/kfserving/server.py]: It was failing for me when I added parent as then both would by default add help methods.

File: python/setup.py (hunk 3 above)
Comment (cliveseldon) [reply python/setup.py]: Is the intention to have separate pypi packages for kfserving and xgbserver as installed packages?
When I run `pip install -e .` it only says its installed xgbserver which I thought was strange though it does install both. Can you point me to the convention on this and how we would push the modules to pypi?

File: python/xgbserver/__main__.py (hunk 4 above)
Comment (ellis-bigelow) [reply python/xgbserver/__main__.py]: Very strange -- I wonder if this is due to versioning. I think it's fine to be explicit.

File: python/setup.py (hunk 3 above)
Comment (ellis-bigelow) [reply python/setup.py]: kfserving and xgbserver are definitely separate packages with separate dependencies. I don't think we actually want to publish the xgbserver at all. Perhaps we can only do setup.py for kfserving.

I hadn't gotten that far ;). It's totally possible I'm doing a pypi anti-pattern, as I've never published a package before. I'll defer to your judgement.

File: python/kfserving/server.py (hunk 1 above)
Comment (ellis-bigelow) [reply python/kfserving/server.py]: Odd. Should this be true then?

Comment (cliveseldon) [reply python/kfserving/server.py]: I think it should be false as otherwise I get an error as both argsparse try to add the same `-h` argument.

File: python/setup.py (hunk 3 above)
Comment (cliveseldon) [reply python/setup.py]: ok - so the xgbserver would just be added to a Dockerfile?

 I'm not sure what is the best way to handle this with the requirements for xgbserver - maybe a requirements_xgbserver.txt or something?

File: python/xgbserver/__main__.py (hunk 4 above)
Comment (cliveseldon) [reply python/xgbserver/__main__.py]: which versioning? I ,eam when I run:

```
python -m xgbserver
usage: __main__.py [-h] [--http_port HTTP_PORT] [--grpc_port GRPC_PORT]
                   [--protocol {tensorflow.http,seldon.http}] --model_dir
                   MODEL_DIR [--model_name MODEL_NAME]
__main__.py: error: the following arguments are required: --model_dir
```

File: python/setup.py (hunk 3 above)
Comment (yuzisun) [reply python/setup.py]: I think the `setup.py` should be in each submodule and  in xgbserver's `setup.py` you can add 
```
install_requires=[
'kfserving>=0.1.0'
]
```

Comment (cliveseldon) [reply python/setup.py]: So like I originally proposed ?
```
   python/kfserving/setup.py
   python/kfserving/kfservng/<python module code>
   python/xgbserver/setup.py
   python/xgbserver/xgbserver/<python module code>
```

We can still just publish to pypi only `kfserving` I suppose

Comment (yuzisun) [reply python/setup.py]: ye this looks good to me.

Comment (ellis-bigelow) [reply python/setup.py]: SGTM. We will separate this into it's own repo soon. I need to coordinate with Abhishek. Any recommendations on name? 
kfserving-core? 
kfservers?

File: python/xgbserver/__main__.py (hunk 4 above)
Comment (ellis-bigelow) [reply python/xgbserver/__main__.py]: That's expected and what I get too. Just point to a model file somewhere. I have some samples in gs://kfserving-samples that we can wire up in a README. I won't be getting much work done this weekend  other than reviewing though.

File: python/setup.py (hunk 3 above)
Comment (animeshsingh) [reply python/setup.py]: i would take the same approach as kubeflow core repo, and the operators for tensorflow/pytorch etc go in their own repos. In that case, this this becomes kfserving, and rest of them can xgboost-serving etc...

File: python/xgbserver/__main__.py (hunk 4 above)
Comment (animeshsingh) [reply python/xgbserver/__main__.py]: @ellis-bigelow it maybe also be good to put strawman dev instructions out there now...

**Comment by cliveseldon** [comment]
I have update the project structure. Perhaps can I suggest:

 * We keep this structure for now and merge. 
 * We discuss moving the code to separate repos at the next project meeting @yuzisun @ellis-bigelow

File: python/xgbserver/setup.py [hunk 5]
```diff
@@ -28,6 +12,7 @@
     python_requires='>3.4',
     packages=find_packages("xgbserver"),
     install_requires=[
+        "kfserver==0.1.0",
         "tornado >= 1.4.1",
```
Comment (ellis-bigelow) [inline python/xgbserver/setup.py]: It doesn't require tornado

File: python/kfserving/kfserving/server.py [hunk 6]
```diff
@@ -11,7 +11,7 @@
 DEFAULT_GRPC_PORT = 8081
 DEFAULT_PROTOCOL = "tensorflow/HTTP"
 
-parser = argparse.ArgumentParser()
+parser = argparse.ArgumentParser(add_help=False)
```
Comment (ellis-bigelow) [inline python/kfserving/kfserving/server.py]: Can we set this to true?

**Comment by ellis-bigelow** [comment]
Thanks Clive!
A couple minor comments. Can you also update your commit message to something more descriptive, e.g. "Refactors xgbserver and kfserver to separate pypi folders"?

@anyone, once these changes are in, feel free to LGTM them

/approve

File: python/kfserving/kfserving/server.py (hunk 6 above)
Comment (cliveseldon) [reply python/kfserving/kfserving/server.py]: This doesn't work for me 
```
python -m xgbserver
Traceback (most recent call last):
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/clive/go/src/github.com/kubeflow/kfserving/python/xgbserver/xgbserver/__main__.py", line 9, in <module>
    parser = argparse.ArgumentParser(parents=[kfserving.server.parser])
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/argparse.py", line 1672, in __init__
    self._add_container_actions(parent)
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/argparse.py", line 1441, in _add_container_actions
    group_map.get(action, self)._add_action(action)
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/argparse.py", line 1571, in _add_action
    action = super(_ArgumentGroup, self)._add_action(action)
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/argparse.py", line 1381, in _add_action
    self._check_conflict(action)
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/argparse.py", line 1520, in _check_conflict
    conflict_handler(action, confl_optionals)
  File "/home/clive/anaconda3/envs/kfserving/lib/python3.7/argparse.py", line 1529, in _handle_conflict_error
    raise ArgumentError(action, message % conflict_string)
argparse.ArgumentError: argument -h/--help: conflicting option strings: -h, --help
```

`False` makes sense to me as we assume child args processors will create the help.
